todo_app/app.py

```diff
@@ -1,57 +1,146 @@
-import requests
-import json
-import os
+from flask_login import login_required, LoginManager
 
 def create_app():
 
+    # App Setup
     config = Config()
     app = Flask(__name__)
     app.config.from_object(config)
     mongo_db = MongoDBClient(config.mongo_db_connection, config.mongo_db_name)
+    # Auth Setup
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    # when user connects - give their cookie a state - a session Variable
+
+    # check the returned state is same as session variable
+
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        state = str(uuid.uuid4())
+        session['state'] = state
+        url = "https://github.com/login/oauth/authorize"
+        headers = {
+            "Accept": "application/json"
+        }
+        params = {
+            "client_id": config.git_client_id,
+            "state" : state
+        }
+        response = requests.request("GET",url,headers=headers,params=params)
+        return redirect(response.url)
+
+    # Come to this route when no cookie present
+    @app.route('/login/callback', methods=['GET'])
+    def authenticate_user():
+        if (not 'state' in session) or (session['state'] != request.args['state']):
+            return render_template('unauthorised.html')
+        if "error" in request.args:
+            return render_template('unauthorised.html')
+        code = request.args['code']
+        url = "https://github.com/login/oauth/access_token"
+        headers = {
+            "Accept": "application/json"
+        }
+        params = {
+            "client_id": config.git_client_id,
+            "client_secret": config.git_client_secret,
+            "code": code
+        }
+        resp = requests.request("POST",url,headers=headers,params=params)
+        if not resp.ok:
+            return render_template('unauthorised.html')
+        access_token = json.loads(resp.text)["access_token"]
+        headers['Authorization'] = "token " + access_token
+        url = "https://api.github.com/user"
+        user_resp = requests.request("GET",url,headers=headers,params=params)
+        if not resp.ok:
+            return render_template('unauthorised.html')
+        user_id = json.loads(user_resp.text)["id"]
+        user = User(user_id)
+        logged_in = flask_login.login_user(user)
+        return redirect(url_for('index'))
+
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
+
+    def write_required(func):
+        def wrapper_write_required(*args, **kwargs):
+            allowed = False
+            cur_user = flask_login.current_user
+            # Check to see if we are in test mode or if we are a writer
+            if config.LOGIN_DISABLED or cur_user.role == "writer":
+                allowed = True            
+            if allowed:
+                return func(*args, **kwargs)
+            else:
+                return render_template('unauthorised.html')
```

Comment:
It would be good to use a 401 Unauthorized response code here too (nice unauthorised page though 😄)

```suggestion
                return render_template('unauthorised.html'), 401
```